src/test/java/org/greenbuttonalliance/gbaresourceserver/usage/repository/IntervalBlockRepositoryTest.java

```diff
@@ -0,0 +1,200 @@
+
+	private static final DateTimeFormatter SQL_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
+
+	@BeforeEach
+	public void initTestData() {
+		intervalBlockRepository.deleteAllInBatch();
+		intervalBlockRepository.saveAll(buildTestData());
+	}
+
+	@Test
+	public void findByPresentId_returnsMatching() {
+		UUID presentUuid = UuidCreator.getNameBasedSha1(UuidCreator.NAMESPACE_URL, PRESENT_SELF_LINK);
+		UUID foundUuid = intervalBlockRepository.findById(presentUuid).map(IntervalBlock::getUuid).orElse(null);
+
+		Assertions.assertEquals(
+			presentUuid,
+			foundUuid,
+			() -> String.format("findById with %s returns entity with ID %s", presentUuid, foundUuid)
+		);
+	}
+
+	@Test
+	public void findByNotPresentId_returnsEmpty() {
+		UUID notPresentUuid = UuidCreator.getNameBasedSha1(UuidCreator.NAMESPACE_URL, NOT_PRESENT_SELF_LINK);
+		Optional<IntervalBlock> intervalBlock = intervalBlockRepository.findById(notPresentUuid);
+
+		Assertions.assertTrue(
+			intervalBlock.isEmpty(),
+			() -> String.format("findById with %s returns entity with ID %s", notPresentUuid, intervalBlock.map(IntervalBlock::getUuid).orElse(null))
+		);
+	}
+
+	@Test
+	public void findAll_returnsAll() {
+		int findByAllSize = intervalBlockRepository.findAll().size();
+		int testDataSize = buildTestData().size();
+
+		Assertions.assertEquals(
+			findByAllSize,
+			testDataSize,
+			() -> String.format("findByAll size of %s does not match test data size of %s", findByAllSize, testDataSize)
+		);
+	}
+
+	@Test
+	public void entityMappings_areNotNull() {
+		IntervalBlock fullyMappedBlock = intervalBlockRepository.findById(UuidCreator.getNameBasedSha1(UuidCreator.NAMESPACE_URL, PRESENT_SELF_LINK)).orElse(null);
+		Assumptions.assumeTrue(fullyMappedBlock != null);
+
+		// build functions to test mappings from IntervalBlock <--> IntervalReading <--> ReadingQuality
+		Function<IntervalBlock, Optional<Set<IntervalReading>>> blockToReadings = ib -> Optional.ofNullable(ib.getIntervalReadings());
+		Function<IntervalBlock, Optional<Set<ReadingQuality>>> blockToReadingQualities = blockToReadings.andThen(opt -> opt.flatMap(
+			readings -> readings.stream().findFirst().map(IntervalReading::getReadingQualities)
+		));
+		Function<IntervalBlock, Optional<IntervalReading>> blockToReadingReversed = blockToReadingQualities.andThen(opt -> opt.flatMap(
+			readingQualities -> readingQualities.stream().findFirst().map(ReadingQuality::getReading)
+		));
+		Function<IntervalBlock, Optional<IntervalBlock>> blockToBlockReversed = blockToReadingReversed.andThen(opt -> opt.map(
+			IntervalReading::getBlock
+		));
+
+		Assertions.assertAll(
+			"Entity mapping failures for block " + fullyMappedBlock.getUuid(),
+			Stream.of(blockToReadings, blockToReadingQualities, blockToReadingReversed, blockToBlockReversed)
+				.map(mappingFunc ->
+					() -> Assertions.assertTrue(mappingFunc.apply(fullyMappedBlock).isPresent()))
+		);
+	}
```

Comment:
This gets a little complicated/messy, but I think testing the JPA cardinality mappings (one-to-one, one-to-many, many-to-many, many-to-many) is quite valuable, as references can unexpectedly turn up `null` if the mapping annotations aren't done carefully.

Reply:
I agree!!